File: TopKElements.java
/*
 * problem link : https://leetcode.com/problems/top-k-frequent-elements/description/
 */


import java.util.*;

public class TopKElements {
        public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        // for (int num : nums) map.put(num, map.getOrDefault(num, 0) + 1);
        for (int num : nums) {
           if (map.containsKey(num)) map.put(num, map.get(num) + 1);
           else map.put(num, 1);
        }
        List<Integer>[] freqs = new List[nums.length + 1];
        for (int i = 0; i < freqs.length; i++) freqs[i] = new ArrayList<>();
        for (Map.Entry<Integer, Integer> pair : map.entrySet()) 
            freqs[pair.getValue()].add(pair.getKey());
        int[] res = new int[k];
        int idx = 0;
        for (int i = freqs.length - 1; i >= 0; i--) {
            for (int num : freqs[i]) {
                res[idx++] = num;
                if (idx == k) return res;
            }
        }
        return res;
    }
}

File: top_k_elements.py
'''
  :problem link : https://neetcode.io/problems/top-k-elements-in-list
'''
from typing import List
from collections import Counter

# time complexity O(nlogn)
def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freqs = Counter(nums)
        res = dict(sorted(freqs.items(), key=lambda x: x[1], reverse=True))
        res = list(res.keys())[:k]
        return res
      
      
# time complexity O(n)
def topKFrequent(self, nums: List[int], k: int) -> List[int]:
    freqs = Counter(nums)
    orderd_freqs = [[] for _ in range(len(nums)+1)]
    for val, idx in freqs.items():
        orderd_freqs[idx].append(val)
    res = []
    for ls in orderd_freqs[::-1]:
        for num in ls:
            res.append(num)
            if len(res) == k: return res